dbr quant overhead[4/x]: speed up hook type calculations

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68432 dbr quant overhead[9/x]: precalculate when to skip op_convert_after_hook
* #68431 dbr quant: remove more asserts from hot paths
* #68404 speed up quantized relu6 inplace kernel
* #68374 dbr quant: simplify relatedness logic
* #68373 dbr quant overhead[8/x]: small speedup in op_needs_quantization
* #68372 dbr quant overhead[7/x]: speed up AutoQuantizationState.reset_to_new_call
* #68371 dbr quant overhead[6/x]: remove unneeded isinstance checks in `op_convert_before_hook`
* #68370 dbr quant overhead[5/x]: remove unnecessary asserts
* #68369 dbr quant: clean up key types in AutoQuantizationState mappings
* **#68351 dbr quant overhead[4/x]: speed up hook type calculations**
* #68350 dbr quant overhead[3/x]: speed up AutoQuantizationState.mark_cur_op_complete

Summary:

Speeds up `get_module_hook_type` and `get_torch_function_hook_type` by
bypassing the expensive `torch.nn.Module` getters and setters and
fetching `_auto_quant_state` directly.

Test plan:

Model level benchmarking is noisy.  Individual `cProfile` results:

```
// MobileNetV2, 1x3x224x224 input, % of time spent by function during DBR convert

// before
get_module_hook_type - 5.96%
get_torch_function_hook_type - 2.24%

// after
get_module_hook_type - 2.10%
get_torch_function_hook_type - 0.57%
```

Differential Revision: [D32463756](https://our.internmc.facebook.com/intern/diff/D32463756)

vkuzo committed:
dbr quant overhead[4/x]: speed up hook type calculations

Summary:

Speeds up `get_module_hook_type` and `get_torch_function_hook_type` by
bypassing the expensive `torch.nn.Module` getters and setters and
fetching `_auto_quant_state` directly.

Test plan:

Model level benchmarking is noisy.  Individual `cProfile` results:

```
// MobileNetV2, 1x3x224x224 input, % of time spent by function during DBR convert

// before
get_module_hook_type - 5.96%
get_torch_function_hook_type - 2.24%

// after
get_module_hook_type - 2.10%
get_torch_function_hook_type - 0.57%
```

[ghstack-poisoned]

vkuzo committed:
Update on "dbr quant overhead[4/x]: speed up hook type calculations"


Summary:

Speeds up `get_module_hook_type` and `get_torch_function_hook_type` by
bypassing the expensive `torch.nn.Module` getters and setters and
fetching `_auto_quant_state` directly.

Test plan:

Model level benchmarking is noisy.  Individual `cProfile` results:

```
// MobileNetV2, 1x3x224x224 input, % of time spent by function during DBR convert

// before
get_module_hook_type - 5.96%
get_torch_function_hook_type - 2.24%

// after
get_module_hook_type - 2.10%
get_torch_function_hook_type - 0.57%
```

[ghstack-poisoned]

vkuzo committed:
Update on "dbr quant overhead[4/x]: speed up hook type calculations"


Summary:

Speeds up `get_module_hook_type` and `get_torch_function_hook_type` by
bypassing the expensive `torch.nn.Module` getters and setters and
fetching `_auto_quant_state` directly.

Test plan:

Model level benchmarking is noisy.  Individual `cProfile` results:

```
// MobileNetV2, 1x3x224x224 input, % of time spent by function during DBR convert

// before
get_module_hook_type - 5.96%
get_torch_function_hook_type - 2.24%

// after
get_module_hook_type - 2.10%
get_torch_function_hook_type - 0.57%
```

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32463756).

jerryzh168 commented on torch/ao/quantization/_dbr/utils.py:
```diff
@@ -498,10 +498,11 @@ def get_torch_function_hook_type(
+        '_auto_quant_state' in parent_module.__dict__['_modules']
     needs_op_hooks = parent_module_has_qstate and \
-        parent_module._auto_quant_state.cur_op_needs_hooks(func)  # type: ignore[union-attr, operator]
+        parent_module.__dict__['_modules']['_auto_quant_state'].cur_op_needs_hooks(func)  # type: ignore[union-attr, operator]
```
wondering should we put write some utility functions ourselves here to make it clearer that we are doing this for performance?

vkuzo replied:
added a comment on the reasoning, not sure if adding one more function would increase the overhead, hope that's ok!

vkuzo committed:
Update on "dbr quant overhead[4/x]: speed up hook type calculations"


Summary:

Speeds up `get_module_hook_type` and `get_torch_function_hook_type` by
bypassing the expensive `torch.nn.Module` getters and setters and
fetching `_auto_quant_state` directly.

Test plan:

Model level benchmarking is noisy.  Individual `cProfile` results:

```
// MobileNetV2, 1x3x224x224 input, % of time spent by function during DBR convert

// before
get_module_hook_type - 5.96%
get_torch_function_hook_type - 2.24%

// after
get_module_hook_type - 2.10%
get_torch_function_hook_type - 0.57%
```

Differential Revision: [D32463756](https://our.internmc.facebook.com/intern/diff/D32463756)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32463756).

vkuzo committed:
Update on "dbr quant overhead[4/x]: speed up hook type calculations"


Summary:

Speeds up `get_module_hook_type` and `get_torch_function_hook_type` by
bypassing the expensive `torch.nn.Module` getters and setters and
fetching `_auto_quant_state` directly.

Test plan:

Model level benchmarking is noisy.  Individual `cProfile` results:

```
// MobileNetV2, 1x3x224x224 input, % of time spent by function during DBR convert

// before
get_module_hook_type - 5.96%
get_torch_function_hook_type - 2.24%

// after
get_module_hook_type - 2.10%
get_torch_function_hook_type - 0.57%
```

Differential Revision: [D32463756](https://our.internmc.facebook.com/intern/diff/D32463756)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32463756).

vkuzo committed:
Update on "dbr quant overhead[4/x]: speed up hook type calculations"


Summary:

Speeds up `get_module_hook_type` and `get_torch_function_hook_type` by
bypassing the expensive `torch.nn.Module` getters and setters and
fetching `_auto_quant_state` directly.

Test plan:

Model level benchmarking is noisy.  Individual `cProfile` results:

```
// MobileNetV2, 1x3x224x224 input, % of time spent by function during DBR convert

// before
get_module_hook_type - 5.96%
get_torch_function_hook_type - 2.24%

// after
get_module_hook_type - 2.10%
get_torch_function_hook_type - 0.57%
```

Differential Revision: [D32463756](https://our.internmc.facebook.com/intern/diff/D32463756)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32463756).

vkuzo committed:
Update on "dbr quant overhead[4/x]: speed up hook type calculations"


Summary:

Speeds up `get_module_hook_type` and `get_torch_function_hook_type` by
bypassing the expensive `torch.nn.Module` getters and setters and
fetching `_auto_quant_state` directly.

Test plan:

Model level benchmarking is noisy.  Individual `cProfile` results:

```
// MobileNetV2, 1x3x224x224 input, % of time spent by function during DBR convert

// before
get_module_hook_type - 5.96%
get_torch_function_hook_type - 2.24%

// after
get_module_hook_type - 2.10%
get_torch_function_hook_type - 0.57%
```

Differential Revision: [D32463756](https://our.internmc.facebook.com/intern/diff/D32463756)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32463756).

vkuzo committed:
Update on "dbr quant overhead[4/x]: speed up hook type calculations"


Summary:

Speeds up `get_module_hook_type` and `get_torch_function_hook_type` by
bypassing the expensive `torch.nn.Module` getters and setters and
fetching `_auto_quant_state` directly.

Test plan:

Model level benchmarking is noisy.  Individual `cProfile` results:

```
// MobileNetV2, 1x3x224x224 input, % of time spent by function during DBR convert

// before
get_module_hook_type - 5.96%
get_torch_function_hook_type - 2.24%

// after
get_module_hook_type - 2.10%
get_torch_function_hook_type - 0.57%
```

Differential Revision: [D32463756](https://our.internmc.facebook.com/intern/diff/D32463756)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32463756).
